### PR TITLE
Feature: add os.sdk sub-settings for Apple

### PR DESCRIPTION
**Pull request by SSE4**
related: #8113 

- add `os.sdk` sub-setting for Apple platforms (iOS/watchOS/tvOS/macOS)
- select correct SDK in `apple_sdk_name` based on `os.sdk`
- return correct `apple_deployment_target_flag` for simulators
- as now we have to accept `os.sdk` in tools, they were changed to accept `ConanFile` as a first argument
- `conan_v2_behavior` issued for old syntax (individual settings values instead of `ConanFile`)

it makes it possible to distinguish ARMv8 iPhone and ARMv8 iPhone Simulator (M1) binaries (they are really different, incompatible binaries).

/cc @theodelrieu 

Changelog: Feature: Add os.sdk sub-settings for Apple
Docs: https://github.com/conan-io/docs/pull/1981

- [x] Refer to the issue that supports this Pull Request.
- [x] If the issue has missing info, explain the purpose/use case/pain/need that covers this Pull Request.
- [x] I've read the [Contributing guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).
- [x] I've followed the PEP8 style guides for Python code.
- [ ] I've opened another PR in the Conan docs repo to the ``develop`` branch, documenting this one. 

<sup>**Note:** By default this PR will skip the slower tests and will use a limited set of python versions. Check [here](https://github.com/conan-io/conan/blob/develop/.github/PR_INCREASE_TESTING.md) how to increase the testing level by writing some tags in the current PR body text.</sup>

### CODE REVIEW

File: conans/test/unittests/util/apple_test.py [hunk 1]
```diff
@@ -44,22 +37,28 @@ def test_to_apple_arch(self):
         self.assertIsNone(tools.to_apple_arch('mips'))
 
     def test_apple_sdk_name(self):
-        self.assertEqual(tools.apple_sdk_name(FakeSettings('Macos', 'x86')), 'macosx')
```
Comment (memsharded) [inline conans/test/unittests/util/apple_test.py]: If the intent (it should be) is not to break existing calls, leave these tests unmodified, and exclusively add new tests.

**Review by memsharded — CHANGES_REQUESTED** [review]
I think it is good, and could be merged once the tests are changed, to leave the old ones, making sure we are not breaking current behavior.

File: conans/test/unittests/util/apple_test.py (hunk 1 above)
Comment (SSE4) [reply conans/test/unittests/util/apple_test.py]: okay, sure, I'll change

**Review by memsharded — COMMENTED** [review]
Also, you could leave ``settings`` as argument for this functionality. You are doing 2 different things in this PR:
- The os.sdk functionality
- An unrelated change to allow passing the conanfile as argument

All tools new v2 mode will be moved to ``conan.tools`` new space, and imported from there explicitly, users will need to change their recipes to use the new tools, conanfile as argument can be enforced there.

**Comment by SSE4** [comment]
@memsharded the reason for the change is that most apple tools accept individual settings values. right now, it turns out it's not enough, and complete settings context is needed.
the process might be done in 2 phases, tools might be changed to accept `settings`, then, at some later point, they might be changed to accept `conanfile` instance. IMO it would be a nightmare for recipe authors to change all recipe(s) twice, first time to accept `settings`, second time to accept `conanfile`. I personally would prefer if I only need to do such a transition once.

File: conans/client/tools/apple.py [hunk 2]
```diff
@@ -50,12 +91,29 @@ def apple_deployment_target_env(os_, os_version):
...
+        os_sdk = conanfile.settings.get_safe("os.sdk")
+        arch = conanfile.settings.get_safe("arch")
+    else:
+        conan_v2_behavior('first argument of "apple_deployment_target_flag" has to be ConanFile,'
+                          'use "tools.apple_deployment_target_flag(self)" instead.')
+
+    if not os_sdk:
+        os_sdk = _guess_apple_sdk_name(os_, arch)
+    if os_sdk:
+        os_sdk = os_sdk.lower()
```
Comment (danimtb) [inline conans/client/tools/apple.py]:
```suggestion
    os_sdk = _guess_apple_sdk_name(os_, arch) if not os_sdk else os_sdk.lower()
```

**Review by danimtb — CHANGES_REQUESTED** [review]
I'd rather not change the apple tools that currently don't need to use the conanfile. In this case, none of them need to change to accept a ConanFile object in order to better support the new `os.sdk` subsetting. IMO, let's just add the new parameters to `apple_deployment_target_flag` and leave the other tools for a future discussion

**Comment by SSE4** [comment]
@memsharded @danimtb I have updated according to your suggestions

**Review by danimtb — APPROVED** [review]

**Comment by SSE4** [comment]
~~@danimtb any idea why `virtualenv` might be MIA?~~

File: conans/client/tools/apple.py [hunk 3]
```diff
@@ -39,6 +35,17 @@ def apple_sdk_name(settings):
...
 
 
+def apple_sdk_name(settings):
+    """returns proper SDK name suitable for OS and architecture
+    we're building for (considering simulators)"""
+    arch = settings.get_safe('arch')
+    os_ = settings.get_safe('os')
+    os_sdk = settings.get_safe('os.sdk')
+    if os_sdk:
+        return os_sdk.lower()
```
Comment (memsharded) [inline conans/client/tools/apple.py]: If in every single place the sdk is used is lowercase, wouldn't it be better to define them as lowercase directly in the settings?

**Review by memsharded — APPROVED** [review]
LGTM, please fix conflict and comment and it can be merged for 1.33.

File: conans/client/tools/apple.py (hunk 3 above)
Comment (SSE4) [reply conans/client/tools/apple.py]: I'd personally prefer always lower-case as it's less error-prone. @theodelrieu  WDYT?

Comment (theodelrieu) [reply conans/client/tools/apple.py]: @SSE4 Agreed, lowercase looks good.

Comment (SSE4) [reply conans/client/tools/apple.py]: changed to lowercase

**Pull request merged**